File: Insert GCDs in Linked List - Leetcode 2807/Insert GCDs in Linked List - Leetcode 2807.py
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = head
        cur = head.next
        
        while cur:
            gcd = math.gcd(cur.val, prev.val) # A
            g = ListNode(gcd)
            prev.next = g
            g.next = cur
            prev = cur
            cur = cur.next

        return head

# Time Complexity: O(n * A)
# Space Complexity: O(1)

File: Insert GCDs in Linked List - Leetcode 2807/Insert GCDs in Linked List - Leetcode 2807.cpp
#include <algorithm> // for std::gcd

class Solution {
public:
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ListNode* prev = head;
        ListNode* cur = head->next;

        while (cur != nullptr) {
            int gcd = std::gcd(cur->val, prev->val); // A
            ListNode* g = new ListNode(gcd);
            prev->next = g;
            g->next = cur;
            prev = cur;
            cur = cur->next;
        }

        return head;
    }
};
